### ISSUE TITLE
`borg create -x` traverses filesystems if parent_fs == child_fs

### ISSUE THREAD

**Issue by bebehei** [borgbackup/borg#2141]
I'm currently migrating towards borgbackup. As always while switching infrastructure, you sometimes hit old yet unknown bugs: For some reasons my root-partition was mounted on `/` and for some reasons on `/mnt/`, too.

While creating a borgbackup archive of `/`, borg traverses the mountpoint of `/mnt`. It's the same filesystem (but it's still a mountpoint!).

Although calling `borg create ` with `-x`-option, both `/` and `/mnt` get archived.

```
[root@nsd:~]$ mount | grep /dev/vda
/dev/vda2 on / type ext4 (rw,relatime,data=ordered)
/dev/vda2 on /mnt type ext4 (rw,relatime,data=ordered)
[root@nsd:~]$ borg create -x steak:nsd::$(date +'%Y-%m-%d-%H:%M') /
[root@nsd ~]$ borg info steak:nsd::2017-02-11-15:27
Name: 2017-02-11-15:27
Fingerprint: c4775b04d5ebf8d81b5243b59616ba191290b5b9663d8392a558250f05d086b2
Hostname: nsd
Username: root
Time (start): Sat, 2017-02-11 15:27:58
Time (end):   Sat, 2017-02-11 15:34:33
Command line: /usr/bin/borg create -x steak:nsd::2017-02-11-15:27 /
Number of files: 132670

                       Original size      Compressed size    Deduplicated size
This archive:                4.85 GB              4.85 GB              2.41 GB
All archives:                4.85 GB              4.85 GB              2.41 GB

                       Unique chunks         Total chunks
Chunk index:                   61225               133578
[root@nsd ~]$ borg list --short steak:nsd::2017-02-11-15:27  | wc -l
160025
[root@nsd ~]$ borg list --short steak:nsd::2017-02-11-15:27 | grep '^mnt' | wc -l
80015
```
So, creating an archive always yields to the double the Original Size but because of borg's deduplication strategy this is just a minor issue.

Also someone has to be as stupid as me and has to mount the backed up volume twice.

But pain starts when extracting the archive, when copying out the doubled size of the harddisk and restoring takes doubled time.

**Comment by enkore** [comment]
I'm not sure if there even is a ~portable way to check for this specific case. I would expect that eg. all the stuff from statvfs() would be the same (since it's the same physical FS) and that the device from stat() would also be the same (same physical device as well). I suspect the only way, if one wanted to fix this, would be to actually enumerate mounted filesystems, which is not portable afaik.

It's a funny situation, where "stay in same file system, do not cross mount points" is both true and false: it stays in the same file system, but did cross over a mount point.

**Comment by textshell** [comment]
Yes. And if we try to enumerate mountpoints (which is not easy in the general case with namespaces and filesystems hiding each other) we should check not to make the same mistakes obnam made so it had to roll back the changes (iirc).

bebehei noted that `find` and `mountpoint` both detect something. In case of `find` is seems to be a general cycle detection (storing all dev/inode tuples and bailing out if one repeats) and `mountpoint` has this very strange special caseing:

```
	r = (st.st_dev != st2.st_dev) ||
	    (st.st_dev == st2.st_dev && st.st_ino == st2.st_ino);
```

**Comment by enkore** [comment]
This would make sense if `st` where the assumed root directory of an FS and `st2` is the suspected mountpoint.

**Comment by textshell** [comment]
May a simple way would be to remember all dev/inode pairs of all parent directories and see is anything repeats. Not sure it is worth it though. Because bind mounts still are likely undetected. And while it is a bit more general than what `mountpoint` does is seems still like quite a special case.

**Comment by enkore** [comment]
Yes. Perhaps just adjust the wording of the --one-file-system option so that it "legally" includes this possibility. Something like "stay in the same file system, do not traverse into other file systems (mountpoints are not stored)". (The latter clarification is something I wanted to put there since #1033)

**Comment by bebehei** [comment]
> bebehei noted that find and mountpoint both detect something.

Researching again: find's algorithm is wrong.

```
[root@buildbot:/] # mount /dev/vda2 /home/asdf
[root@buildbot:/] # find /home -mount | grep /root file
find: File system loop detected; '/home/asdf/home' is part of the same file system loop as '/home'.
[...]
/home/asdf/root/.ssh
[...]
```

find only won't traverse into the same subdirectory of this filesystem, but it will include the rest of the FS.

Today I got heavily disappointed of the `-x`, `-mount` and `--one-filesystem` options in common Unix programs.

> Yes. Perhaps just adjust the wording of the --one-file-system option so that it "legally" includes this possibility.

@enkore Yes, this is the best way to "fix" it.

**Comment by enkore** [comment]
"It's not their fault" -- the semantics that you can mount the same FS twice, or have bind mounts, or mount subvolumes (btrfs, ZFS) are all quite new and were not present when most of these were written... or the portable-ish APIs meant to deal with it (stat, statvfs) :)

**Comment by ThomasWaldmann** [comment]
so, just to clarify this: "bug" "documentation" means a bug in the documentation (not a code bug we have to fix plus a documentation issue).

**Comment by enkore** [comment]
Fixed by #2146
